### PR TITLE
MF-1830 - Switch to Semaphore 2.0

### LINKED ISSUE #1830: Switch to Semaphore 2.0
**ENHANCEMENT**
1. Describe the enhancement you are requesting.
Semaphore CI is sunsetting its Semaphore Classic tool: https://semaphoreci.com/blog/semaphore-classic-deprecation
Since it provides better value than GitHub Actions for the Mainflux project, we will switch to Semaphore 2.0.

3. Indicate the importance of this enhancement to you (must-have, should-have, nice-to-have).
This is a must-have, since from September our CI will be broken if we don't do this.

### PR DESCRIPTION
**Pull request by WashingtonKK**
### What does this do?
Implement CI using Semaphore 2

### Which issue(s) does this PR fix/relate to?
Resolves #1830.

### List any changes that modify/break current functionality
- CI runs tests only on modified modules

### Have you included tests for your changes?
N/A

### Did you document any new/modified functionality?
Yes

### Notes
N/A

### CODE REVIEW

File: .semaphore/semaphore.yml [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                sudo rm -rf /usr/local/bin/go
+                sudo rm -rf /usr/local/golang
+                sudo rm -rf $GOROOT $GOPATH $GOBIN
+                wget https://go.dev/dl/go$GO_VERSION.linux-amd64.tar.gz
+                sudo tar -C /usr/local -xzf go$GO_VERSION.linux-amd64.tar.gz
+                export GOROOT=/usr/local/go
+                export PATH=$PATH:/usr/local/go/bin
+              fi
+            - export GOBIN=$HOME/go/bin
+            - 'export PATH=$PATH:$GOBIN'
```
Comment (dborovcanin) [inline .semaphore/semaphore.yml]: We probably do not need to update Go anymore. This was only used as a way of enabling the latest version of Go, but backward compatibility guarantees that it should be just fine if the CI Go version is 1 or 2 (but not more) older than the version in `go.mod`.

File: .semaphore/semaphore.yml [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+              MODIFIED_MODULES=$(go list -e -f '{{.Module.Path}}' $(dirname $MODIFIED_FILES) | sort -u)
+
+              # Run tests for modified modules
+              for module in $MODIFIED_MODULES; do
+                go test -mod=vendor -v -race -tags test -coverprofile=profile.out -covermode=atomic "$module/..."
+                if [ -f profile.out ]; then
+                  cat profile.out >> coverage.txt
+                  rm profile.out
+                fi
+              done
```
Comment (dborovcanin) [inline .semaphore/semaphore.yml]: We have only one module. Also, detecting changes for testing is more complex than checking changed files, since it depends on **which files are changed**: if we change access control, for example, we need to test messaging, as well.

**Review by dborovcanin — CHANGES_REQUESTED** [review]
@WashingtonKK Please check examples here: https://semaphoreci.com/blog/revving-up-continuous-integration-with-parallel-testing#h-monorepos.

File: .semaphore/semaphore.yml [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+      when: change_in('/bootstrap/')
+    task:
+      jobs:
+        - commands:
+            - checkout
+            - cd bootstrap
+            - go test -v -coverprofile=coverage.out ./...
+            - go tool cover -html=coverage.out -o coverage.html 
+          name: Test Bootstrap
+  - name: Test certs
```
Comment (rodneyosodo) [inline .semaphore/semaphore.yml]: add tests for `cli`, `coap`, `consumers`, whole `http` sub package not only `/api`, and others

File: .semaphore/semaphore.yml [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+      when: change_in('/logger/')
+    task:
+      jobs:
+        - commands:
+            - checkout
+            - cd logger
+            - go test -v -coverprofile=coverage.out ./...
+            - go tool cover -html=coverage.out -o coverage.html 
+          name: Logger test
+  - name: Test all
```
Comment (rodneyosodo) [inline .semaphore/semaphore.yml]: This doesn't seem to be correct. I'd suggest to test everything probably during merge to main or after a reviewer has commented `LGTM`

**Review by rodneyosodo — CHANGES_REQUESTED** [review]
Add test for building binaries

File: .semaphore/semaphore.yml (hunk 3 above)
Comment (WashingtonKK) [reply .semaphore/semaphore.yml]: On it

**Review by rodneyosodo — CHANGES_REQUESTED** [review]
Add [auto-cancel on new push ](https://docs.semaphoreci.com/reference/pipeline-yaml-reference/#auto_cancel)

File: .semaphore/semaphore.yml [hunk 5]
```diff
@@ -74,12 +74,7 @@ blocks:
             - checkout
             - make test
         run:
-          when:
-            event:
-              only: main
-            changeset:
-              comment:
-                matches: "LGTM"
+          when: "branch" = "master" OR "comment" = "LGTM"
```
Comment (rodneyosodo) [inline .semaphore/semaphore.yml]: It doesn't work for a comment - https://docs.semaphoreci.com/reference/conditions-reference/

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: .semaphore/semaphore.yml (hunk 3 above)
Comment (drasko) [reply .semaphore/semaphore.yml]: Stay consistent and use capital letter for the name of each service (Certs, Things, ...)

**Review by drasko — CHANGES_REQUESTED** [review]
Capitalize service names

File: .semaphore/semaphore.yml [hunk 6]
```diff
@@ -0,0 +1,246 @@
+version: v1.0
+name: CI Pipeline
```
Comment (rodneyosodo) [inline .semaphore/semaphore.yml]: The previous CI run took `23.20m` yet the current CI runs on the master take around `24m`. Shouldn't we be getting faster times?

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: .semaphore/semaphore.yml (hunk 6 above)
Comment (WashingtonKK) [reply .semaphore/semaphore.yml]: Only 3 processes are being run at any instance on the free tier which slows down the whole process. If we had all parallel processes running at the same time concurrently we would have faster times of ~19 min.

File: cmd/opcua/main.go [hunk 7]
```diff
@@ -126,6 +126,7 @@ func main() {
 	if err := env.Parse(&httpServerConfig, env.Options{Prefix: envPrefixHttp, AltPrefix: envPrefix}); err != nil {
 		logger.Fatal(fmt.Sprintf("failed to load %s HTTP server configuration : %s", svcName, err))
 	}
+
```
Comment (SammyOina) [inline cmd/opcua/main.go]: this change is unnecessary

File: docker/.env [hunk 8]
```diff
@@ -61,6 +61,7 @@ MF_USERS_RESET_PWD_TEMPLATE=users.tmpl
 MF_USERS_PASS_REGEX=^.{8,}$$
 MF_USERS_INSTANCE_ID=
 
+
```
Comment (SammyOina) [inline docker/.env]: same here

File: http/api/transport.go [hunk 9]
```diff
@@ -6,6 +6,7 @@ package api
 import (
 	"context"
 	"encoding/json"
+
```
Comment (SammyOina) [inline http/api/transport.go]: here too

**Review by SammyOina — CHANGES_REQUESTED** [review]
remove changes not required in this PR

File: http/api/transport.go (hunk 9 above)
Comment (SammyOina) [reply http/api/transport.go]: probably fixed by rebasing

Comment (WashingtonKK) [reply http/api/transport.go]: The change was associated with a merge from master into the branch to keep it up to date.

File: .semaphore/semaphore.yml [hunk 10]
```diff
@@ -0,0 +1,430 @@
+version: v1.0
```
Comment (rodneyosodo) [inline .semaphore/semaphore.yml]: Remove previous CI

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: .semaphore/semaphore.yml (hunk 4 above)
Comment (dborovcanin) [reply .semaphore/semaphore.yml]: There is probably no need for this test at all, even in cases we merge to `master`. However, we need to be sure we tun tests when it is relevant (e.g. when proto files change, we need to run almost all the tests).

File: .semaphore/semaphore.yml [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+    run:
+      when: "change_in('/ws/', {pipeline_file: 'ignore'})"
+    task:
+      jobs:
+        - commands:
+            - make ws
+            - docker push mainflux.ws
+          name: Push Ws
+
+  - name: Lint and Test CoAp
```
Comment (dborovcanin) [inline .semaphore/semaphore.yml]: Fix typo: `CoAP`. Check for other typos.

File: .semaphore/semaphore.yml [hunk 12]
```diff
@@ -0,0 +1,452 @@
...
+            - ./codecov                 
+  - name: Push Things
+    dependencies:
+      - Lint and Test Things
+    run:
+      when: "change_in('/things', {pipeline_file: 'ignore'})"
+    task:
+      jobs:
+        - commands:
+            - make things
```
Comment (dborovcanin) [inline .semaphore/semaphore.yml]: We need to build a Docker image, not a binary.

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Please add checks for proto file changes, and also connect changes in Things auth with protocol adapters (i.e. adapters must be tested when auth is changed, but artifacts **do not need to be published** since the code actually did not change). Since all Lint and Test XYZ have the same pattern, check if we can use some kind of matrix and template hee.

File: .semaphore/semaphore.yml [hunk 13]
```diff
@@ -0,0 +1,849 @@
...
+            - make -j$NPROC latest
+
+  - name: Compile Check For RabbitMQ
+    dependencies:
+      - Setup
+    task:
+      jobs:
+        - name: Compile For RabbitMQ
+          commands:
+            - MF_BROKER_TYPE=rabbitmq make mqtt
```
Comment (rodneyosodo) [inline .semaphore/semaphore.yml]: Add empty line

Comment (ianmuchyri) [reply .semaphore/semaphore.yml]: @WashingtonKK this is still not addressed

File: http/api/transport.go [hunk 14]
```diff
@@ -78,6 +78,7 @@ func parseSubtopic(subtopic string) (string, error) {
 	}
 	subtopic = strings.ReplaceAll(subtopic, "/", ".")
 
+
```
Comment (ianmuchyri) [inline http/api/transport.go]: unnecessary change

**Comment by SammyOina** [comment]
image
This and other similar commits should not be included, use `git rebase -i master`

File: .semaphore/semaphore.yml [hunk 15]
```diff
@@ -0,0 +1,892 @@
...
+
+                go install google.golang.org/protobuf/cmd/protoc-gen-go@$PROTOC_GEN_VERSION
+                go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@$PROTOC_GRPC_VERSION
+                
+                export PATH=$PATH:/usr/local/bin/protoc
+                export PATH=$PATH:$HOME/go/bin
+              }
+            - setup_protoc
+            - |
+              check_files() {
```
Comment (dborovcanin) [inline .semaphore/semaphore.yml]: Same here.

File: .semaphore/semaphore.yml [hunk 16]
```diff
@@ -0,0 +1,892 @@
...
+  - name: Check Generated Protocol Buffer Files
+    dependencies: []
+    run:
+      when: "change_in(['things/policies/auth.proto', 'users/policies/auth.proto', 'pkg/messaging/message.proto'])"
+    task:
+      jobs:
+        - name: Check Generated Protocol Buffer Files
+          commands:
+            - |
+              setup_protoc() {
```
Comment (dborovcanin) [inline .semaphore/semaphore.yml]: No need to use functions and call them - you can simply exec this line-by-line.

File: .semaphore/semaphore.yml [hunk 17]
```diff
@@ -0,0 +1,892 @@
...
+            - go test ./... --race -v -coverprofile=coverage.out
+            - go tool cover -html=coverage.out -o coverage.html
+            - cache restore codecov
+            - ./codecov
+      secrets:
+        - name: CODECOV_TOKEN
+  - name: Things Auth Changes
+    dependencies: []
+    run:
+      when: "change_in('things/policies/auth.proto')"
```
Comment (dborovcanin) [inline .semaphore/semaphore.yml]: When changes are in Protobuf files, we need all the tests (i.e. add `things/policies/auth.proto` to a `change_in` list for all adapters.

Comment (dborovcanin) [reply .semaphore/semaphore.yml]: Instead of when: `"change_in('things/policies/auth.proto')"` we need to check for changes in **implementation logic of things** - something like `when: "change_in(['things/policies/postgres', 'things/policies/service.go'])"`. Also, no need to run coverage in this block since we did not change the actual code of adapters, so no coverage changes will be present.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: .semaphore/semaphore.yml [hunk 18]
```diff
@@ -0,0 +1,871 @@
...
+  - name: Things Auth Changes
+    dependencies: []
+    run:
+      when: "change_in(['things/policies/postgres', 'things/policies/service.go'])"
+    task:
+      jobs:
+        - name: Test Things
+          commands:
+            - cd things
+            - go test ./... --race
```
Comment (dborovcanin) [inline .semaphore/semaphore.yml]: No need for this, it will be triggered by Thing changes.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: .semaphore/semaphore.yml [hunk 19]
```diff
@@ -0,0 +1,871 @@
...
+  - name: Push Docker Images
+    dependencies:
+      - Setup
+    run:
+      when: branch = 'master'
+    task:
+      jobs:
+        - name: Push Docker Images
+          commands:
+            - make -j$NPROC latest
```
Comment (dborovcanin) [inline .semaphore/semaphore.yml]: No need for this, either. Move publishing to per-service in Build blocks.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: .semaphore/semaphore.yml [hunk 20]
```diff
@@ -850,16 +846,16 @@ blocks:
...
+  # - name: Push Docker Images
+  #   dependencies:
+  #     - Setup
+  #   run:
+  #     when: branch = 'master'
+  #   task:
+  #     jobs:
+  #       - name: Push Docker Images
+  #         commands:
+  #           - make -j$NPROC latest
```
Comment (dborovcanin) [inline .semaphore/semaphore.yml]: Remove commented code.

File: .semaphore/semaphore.yml [hunk 21]
```diff
@@ -0,0 +1,725 @@
...
+      jobs:
+        - name: Build Users
+          commands:
+            - make docker_users
+      secrets:
+        - name: DOCKER_TOKEN            
+        - name: DOCKER_USERNAME                    
+  - name: Lint and Test Things
+    run:
+      when: "change_in(['things', 'cmd/things', 'users/policies/auth.proto'])"
```
Comment (dborovcanin) [inline .semaphore/semaphore.yml]: Let's test if changes are in `/users/policies/postgres/policies.go` and `/users/policies/api/grpc/client.go`, not `user/policies/auth.proto`.

File: .semaphore/semaphore.yml [hunk 22]
```diff
@@ -0,0 +1,725 @@
...
+      jobs:
+        - name: Build Things
+          commands:
+            - make docker_things
+      secrets:
+        - name: DOCKER_TOKEN            
+        - name: DOCKER_USERNAME              
+  - name: Lint and Test CoAP
+    run:
+      when: "change_in(['coap', 'cmd/coap', 'pkg/messaging/message.proto', 'things/policies/postgres', 'things/policies/service.go'])"
```
Comment (dborovcanin) [inline .semaphore/semaphore.yml]:
```suggestion
      when: "change_in(['coap', 'cmd/coap',  'things/policies/postgres/policies.go', 'things/policies/api/grpc/client.go', 'things/policies/service.go'])"
```

Comment (dborovcanin) [reply .semaphore/semaphore.yml]: Do the same for other adapters and services using Things gRPC API.

Comment (dborovcanin) [reply .semaphore/semaphore.yml]: The same pattern goes for services that use Users gRPC API (`'users/policies/postgres/policies.go', 'users/policies/api/grpc/client.go', 'users/policies/service.go'`) so update those, as well.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — CHANGES_REQUESTED** [review]
@WashingtonKK  You are missing these:
```
- echo $DOCKER_TOKEN | docker login --username "$DOCKER_USERNAME" --password-stdin
- docker push mainflux/http:latest
 ```
Ofc, this is for HTTP and we need for others as well.

**Review by dborovcanin — APPROVED** [review]

**Review by arvindh123 — APPROVED** [review]

**Review by drasko — APPROVED** [review]
LGTM
